zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ListenerContainer.java

```diff
@@ -0,0 +1,186 @@
+
+  /**
+   * Add listener to the container with specified key.
+   * @param key the key to register to
+   * @param listener the listener to register
+   * @param persistListener true if the listener is persistent
+   */
+  public void addListener(String key, T listener, boolean persistListener) {
+    addListener(key, listener, persistListener ? _persistentListener : _onetimeListener);
+  }
+
+  /**
+   * Remove listener from the container.
+   * This operation removes both one-time and persistent listener from the specified key.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   */
+  public void removeListener(String key, T listener) {
+    _lock.lock();
+    try {
+      removeFromListenerMap(key, listener, _persistentListener);
+      removeFromListenerMap(key, listener, _onetimeListener);
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Remove listener from the container.
+   * @param key the key to remove
+   * @param listener the listener to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public void removeListener(String key, T listener, boolean persistListener) {
+    _lock.lock();
+    try {
+      if (persistListener) {
+        removeFromListenerMap(key, listener, _persistentListener);
+      } else {
+        removeFromListenerMap(key, listener, _onetimeListener);
+      }
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Remove listeners from the container.
+   * @param key the key to remove
+   * @param listeners the listeners to remove
+   * @param persistListener true if remove the persistent listener, otherwise remove the one-time listener
+   */
+  public void removeListeners(String key, Collection<T> listeners, boolean persistListener) {
+    _lock.lock();
+    try {
+      if (persistListener) {
+        removeFromListenerMap(key, listeners, _persistentListener);
+      } else {
+        removeFromListenerMap(key, listeners, _onetimeListener);
+      }
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  /**
+   * Consume the listeners registered to the given key. The given consumer is applied to all one-time and persistent
+   * listeners with the given key.
+   * The one-time listeners are removed after this operation, as a result, this operation is NOT idempotent.
+   * @param key the key of the listeners
+   * @param consumer the consuming action for the listeners
+   */
+  public void consumeListeners(String key, Consumer<? super T> consumer) {
+    Set<T> onetimeListeners;
+    Set<T> persistentListeners;
+    _lock.lock();
+    try {
+      // remove one-time listeners
+      onetimeListeners = _onetimeListener.remove(key);
+    } finally {
+      _lock.unlock();
+    }
+    if (onetimeListeners == null) {
+      onetimeListeners = Collections.emptySet();
+    }
+    persistentListeners = _persistentListener.getOrDefault(key, Collections.emptySet());
```

Comment:
just curious, why do we need lock for onetimelisteners and not here? Also we can just use getOrDefault for onetimelisteners as well right?

Reply:
For one-time listener, it's a write operation which removes the entry, we'll need lock. For persistent listener, it's pure read, we are able to go lock-free with the specified thread-safe primitives.

Reply:
Oops i missed that part!